keps/0024-parameter-enhancement.md

```diff
@@ -0,0 +1,146 @@
+authors:
+  - "@nfnt"
+owners:
+  - "@nfnt"
+creation-date: 2020-02-21
+last-updated: 2020-02-21
+status: provisional
+---
+
+# Enhanced Operator Parameters
+
+## Table of Contents
+
+- [Enhanced Operator Parameters](#enhanced-operator-parameters)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Support Optional Parameter Types](#support-optional-parameter-types)
+    - [Provide Parameter Validation](#provide-parameter-validation)
+    - [Alternative Considered: Convert YAML in string to dictionaries](#alternative-considered-convert-yaml-in-string-to-dictionaries)
+    - [Alternative Considered: Allow Arbitrary Parameter Values](#alternative-considered-allow-arbitrary-parameter-values)
+  - [Implementation History](#implementation-history)
+
+## Summary
+
+This KEP aims to improve the use-cases that can be covered by parameters. By supporting additional types, templates can access lists and dictionaries. Furthermore, validation can be implemented for known types.
+
+## Motivation
+
+Currently only string parameters are supported by KUDO. This limits the possible use-cases of parameters. Various use cases benefit from having parameters that are lists or dictionaries. While some of this functionality can be implemented in Go templates using [Sprig functions][1], this is unintuitive.
+Having parameter types creates the possibility for parameter validation.
+
+### Goals
+
+- Provide operator developers with more versatile parameters
+- Allow the definition of validation functions for parameters
+
+### Non-Goals
+
+- Change how parameters are defined
+
+## Proposal
+
+### Support Optional Parameter Types
+
+By adding an optional `type` field to the list of parameters, parameter types can be specified. If this field isn't present, it defaults to the `string` type. Other types like `list`, `dict` or `integer` are possible. KUDO's renderer uses the type information to convert the string value to the respective type. For example, for the `dict` type, the input value would be YAML that is unmarshalled to a `map[string]interface{}`. Go's template engine allows to access and iterate list and dictionary items.
+
+Consider a cluster that has nodes in multiple regions. We want to ensure that a nginx deployment has 2 replicas in one region and 1 replica in the other one. Furthermore they should support both HTTP as well as HTTPS, hence need a list of port definitions. These parameters are described by `list` and `dict` types:
+
+```yaml
+parameters:
+  - name: ports
+    description: Ports
+    type: list
+    default: [ 80, 443 ]
+  - name: topology
+    description: Node topology
+    type: dict
+    default: |
+      - region: us-east-1:
+        replicas: 2
+      - region: us-west-1:
+        replicas: 1
+```
+
+The deployment template then iterates over the items of the respective parameters:
```

Comment:
It would be useful to see an example of how easy/complicated it would be to expand a deeply nested object like a `nodeSelector` or `nodeAffinity`? Looking at the simple example below I'm afraid that something deeper might require the template to be very verbose in comparison to what Helm's `toYaml` does.

Reply:
Agreed, it will become verbose. A `toYaml` function is very much needed but not part of this KEP.

Reply:
OK, let me draft another one to build on this.

Reply:
Created PR https://github.com/kudobuilder/kudo/pull/1363